### ISSUE TITLE
Getting muffled voice when using with MPS

### ISSUE THREAD

**Issue by utkarshshukla2912** [k2-fsa/OmniVoice#8]

**Comment by FreedomCoder-dev** [comment]
I ran into the same problem and did some investigation to isolate the root cause. Here's what I found.

**Environment**
- macOS (Apple Silicon, arm64)
- Python 3.12, `omnivoice==0.1.1`, `torch==2.8.0`, `torchaudio==2.8.0`
- Device: `mps` with `dtype=torch.float32`

**Symptom**
When running on MPS, the generated audio is a ~50ms fragment repeated for the entire duration — no intelligible speech. The same text/instruct/config on CPU produces correct, natural-sounding speech.

**Root cause analysis**

I isolated the problem to the **LLM token generation step** (`_generate_iterative`), not the audio tokenizer decoder:

1. **The audio tokenizer decoder is not at fault.** It already runs on CPU (line 264–266 of `omnivoice.py` explicitly forces `tokenizer_device = "cpu"` when the model device is MPS). Feeding MPS-generated tokens into a separately loaded CPU decoder produces the same broken waveform — confirming the tokens themselves are bad.

2. **MPS-generated tokens are numerically diverse but semantically broken.** Comparing MPS vs CPU token outputs for the same input:
   - Both produce 8×87 token grids with ~260–270 unique token IDs
   - No exact column-level repetition in either (checked periods 1–43 and multi-column windows of 3, 5, 7)
   - Token overlap between MPS and CPU is only ~8.7%
   - But the MPS tokens decode to a waveform with near-constant RMS per 50ms chunk (~0.096 ± 0.003), while CPU tokens decode to natural variation (0.05–0.19)

3. **The iterative unmasked decoding loop** (`_generate_iterative`) uses operations including `torch.topk`, `F.log_softmax`, classifier-free guidance arithmetic, and 4D attention masks — all running on MPS. The MPS backend has known silent-correctness issues with batched inference and certain tensor operations (see [pytorch/pytorch#170837](https://github.com/pytorch/pytorch/issues/170837), [pytorch/pytorch#169342](https://github.com/pytorch/pytorch/issues/169342), [pytorch/pytorch#175191](https://github.com/pytorch/pytorch/issues/175191)).

**Workaround**

Use `device_map="cpu"` instead of `device_map="mps"`:

```python
model = OmniVoice.from_pretrained("k2-fsa/OmniVoice", device_map="cpu", dtype=torch.float32)
```

Or via CLI:

```bash
omnivoice-infer --model k2-fsa/OmniVoice --device cpu --text "Hello world" --output out.wav
```

CPU inference on Apple Silicon M-series takes ~20–25s for a ~7s clip, which is acceptable for most use cases.

**Comment by edwios** [comment]
This is NOT a workaround, it is a fallback. mps support simply is broken. Good findings though, at least there is a direction to solve the issue.

**Comment by takuma104** [comment]
I'm also encountering this issue. After seeking help from Claude Code Opus4.6, it appears there may be a numerical precision problem with the MPS attention implementation (SDPA), causing compatibility issues. Setting attn_implementation="eager" as shown below results in slightly slower performance, but seems acceptable compared to using `device_map="cpu"`.

```py
from omnivoice import OmniVoice
import torch

model_eager = OmniVoice.from_pretrained(
    "k2-fsa/OmniVoice",
    device_map="mps",
    dtype=torch.float16,
    attn_implementation="eager", # added to workadound
)
```

**Comment by TBD** [comment]
Claude Code investigation that makes MPS work (also use longer text input - "hello" fails):

> Bug: MPS (Apple Silicon) doesn't gracefully handle attention rows that are
>   entirely False (masked). The uncond CFG batch items have padding positions
>   beyond u_len with all-zero attention rows, causing softmax(-inf, -inf, ...) =
>   NaN. This NaN propagates through the LLM layers and corrupts the valid logits,
>    so argmax always returns 0 — every token becomes token ID 0, producing
>   periodic noise from the codec.
> 
>   Fix (omnivoice/models/omnivoice.py, _generate_iterative): give each padding
>   position a self-attend diagonal entry so no attention row is entirely False:
> 
>   if max_c_len > u_len:
>       pad_diag = torch.arange(u_len, max_c_len, device=self.device)
>       batch_attention_mask[B + i, :, pad_diag, pad_diag] = True

**Comment by duongtugiang** [comment]
which line to apply the patch?

**Comment by zhu-han** [comment]
Thanks for reporting this issue, and a special thanks to @TBD and @Talpik for providing a solution. Could I get more feedback from you guys on whether this fix addresses the issue on your devices?

I’m using a MacBook Pro with Apple M3 Pro. On my device, the original implementation works well with MPS, so I guess this issue only occurs on certain types of devices.

**Comment by TBD** [comment]
@zhu-han same device (macbook m3 pro)

for me it works - i can clone and design voices that sounds good even in 8 steps (around 6s for a short sentence). I am looking in caching intermediate steps in voice cloning and streaming with word highlighting

**Comment by zhu-han** [comment]
> [@zhu-han](https://github.com/zhu-han) same device (macbook m3 pro)
> 
> for me it works - i can clone and design voices that sounds good even in 8 steps (around 6s for a short sentence). I am looking in caching intermediate steps in voice cloning and streaming with word highlighting

Thanks for the feedback! Just to confirm: did you encounter this issue on a MacBook M3 Pro with MPS using the current code, and resolve it with the code you mentioned? If so, could you provide more details? Such as your macOS version, Python version, and whether you were using the same PyTorch version (2.8.0)? I’d like to identify the cause of this issue.

**Comment by TBD** [comment]
# OmniVoice MPS / Voice Design Fixes

## Environment

- **macOS:** 26.3.1 (Tahoe)
- **Python:** 3.12.7
- **PyTorch:** 2.8.0 (macOS); 2.8.0+cu128 (Linux/Win)

## Patch

```diff
diff --git a/omnivoice/models/omnivoice.py b/omnivoice/models/omnivoice.py
index 91d841e..e3439e3 100644
--- a/omnivoice/models/omnivoice.py
+++ b/omnivoice/models/omnivoice.py
@@ -1055,8 +1055,12 @@ class OmniVoice(PreTrainedModel):
 
         # Build style tokens: <|denoise|> + <|lang_start|>...<|lang_end|>
         #                      + <|instruct_start|>...<|instruct_end|>
+        # <|denoise|> is only used when reference audio is provided (voice clone /
+        # denoising mode). During training it is added only for samples with
+        # clean_start_token_idx (i.e. denoising samples). For voice design (no
+        # ref audio) the model was never trained with this token, so omit it.
         style_text = ""
-        if denoise:
+        if denoise and ref_audio_tokens is not None:
             style_text += "<|denoise|>"
         lang_str = lang if lang else "None"
         instruct_str = instruct if instruct else "None"
@@ -1173,6 +1177,12 @@ class OmniVoice(PreTrainedModel):
             batch_input_ids[B + i, :, :u_len] = inp["input_ids"][..., -u_len:]
             batch_audio_mask[B + i, :u_len] = inp["audio_mask"][..., -u_len:]
             batch_attention_mask[B + i, :, :u_len, :u_len] = True
+            # Padding positions beyond u_len have all-False attention rows, which
+            # causes softmax(-inf...) = NaN on MPS. Give each padding position a
+            # self-attend entry so no row is entirely False.
+            if max_c_len > u_len:
+                pad_diag = torch.arange(u_len, max_c_len, device=self.device)
+                batch_attention_mask[B + i, :, pad_diag, pad_diag] = True
 
         tokens = torch.full(
             (B, self.config.num_audio_codebook, max(task.target_lens)),
```

## Changes

1. **`<|denoise|>` token guard** — only emitted when `ref_audio_tokens is not None`. The model was never trained with this token in voice design mode (no ref audio), so adding it unconditionally on `denoise=True` produced garbage output.

2. **MPS NaN fix** — padding rows in the 4-D attention mask beyond `u_len` were all-False, causing `softmax(-inf, ...) = NaN` on Apple MPS. Fixed by setting the diagonal (self-attend) to `True` for every padding position.

**Comment by Conghdos** [comment]
Tôi gặp vấn đề tương tự trên Mac M2 Max

**Comment by zhu-han** [comment]
Thanks. I've merged PR https://github.com/k2-fsa/OmniVoice/pull/13 to address this issue. If anyone encounters a similar problem with the latest code, feel free to reopen this issue.
